pkg/util/resource/monitor.go

```diff
@@ -0,0 +1,236 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"runtime"
+	"runtime/metrics"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/procfs"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/configs"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type ExhaustedError struct{}
+
+func (e *ExhaustedError) Error() string {
+	return "resource exhausted"
+}
+
+type UnsupportedOSError struct{}
+
+func (e *UnsupportedOSError) Error() string {
+	return "resource scanner is only supported in linux"
+}
+
+const heapMetricName = "/memory/classes/Heap/objects:bytes"
+const monitorInterval = time.Second
+const dataPointsToAvg = 30
+
+type IScanner interface {
+	Scan() (Stats, error)
+}
+
+type Scanner struct {
+	proc          procfs.Proc
+	metricSamples []metrics.Sample
+}
+
+type Stats struct {
+	CPU  float64
+	Heap uint64
+}
+
+func NewScanner() (*Scanner, error) {
+	if runtime.GOOS != "linux" {
+		return nil, &UnsupportedOSError{}
```

Comment:
Instead of returning unsupported OS, is it better to create a dummy scanner instead? If we can make scanner an interface it should be easy to create a scanner that does noop.

Reply:
Just saw that we handle the unsupported OS error outside. But we eventually still return error. Can we try to not return error in this case? Not that user friendly for Mac developers.

Reply:
My only concern with noop scanner is the value it returns on `Scan()`. If the noop scanner returns zero-values, how does the resource monitor (or the user) know whether the scanner is actually reporting 0 usage or it's due to unsupported OS?

Reply:
I think it is ok for resource monitor to be not supported in non Linux system and resource usage detected is 0 + some warning log. Otherwise, it is not breaking changes for any users that run Cortex in non Linux systems because they are unable to run it anymore. We do publish non Linux binaries https://github.com/cortexproject/cortex/releases so better to not break this.

Reply:
A common pattern for a feature to support different OS/CPU arch is to use build tag or use different file name for different OS. You can take a look at https://github.com/prometheus/prometheus/tree/701d13abf9fc6fc8ab7ebfb0c112102f8de66529/tsdb/fileutil.

I think it is a better practice than checking `runtime.GOOS` here.